### DISCUSSION TITLE
Struct equivalent of COLUMNS(* REPLACE / EXCLUDE / REGEXP / LAMBDA)

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#13840] [Ideas]
It would be convenient if everything that can be done using star expressions could also be done using regular methods on structs. For example, on a table `df` with a struct column `s` one could then do

```sql
FROM 
  df 
SELECT 
  s.exclude('a'),
  s.replace(a:=b + 1),
  s.select('a', b'), 
  s.select(<lambda>),
  s.select_regexp('<regexp>'),
  s.select_regexp('<regexp>', '<renameexpression>'),
  s.*, --- already exists and works
  s.* + 1, --- doesn't currently work: Binder Error: STAR expression is only allowed as the root element of an expression. Use COLUMNS(*) instead.
  greatest(*s) 
```

In fact, it would feel more structured and less magical to me if
```sql
FROM
  df
SELECT 
  COLUMNS(*),
  COLUMNS(* EXCLUDE ...),
  COLUMNS(* REPLACE ...),
  ...
```

were only (more verbose!) syntactic sugar for 

```sql
FROM 
  df
SELECT 
  df.*,
  df.exclude(...).*,
  df.replace(...).*,
  ...
```

(using the fact that the table name itself provides a struct containing all columns).

This would also allow for things that are not currently available on star expressions, but are frequently requested (see https://github.com/duckdb/duckdb/discussions/13680 , https://github.com/duckdb/duckdb/discussions/6706 , https://github.com/duckdb/duckdb/discussions/13675 , https://github.com/duckdb/duckdb/discussions/13454), such as

```sql
FROM
  df
SELECT
  df.exclude('a', 'b'),
  df.exclude(2, 4),
  df.exclude_regexp('<regexp>'),
  df.reverse(),
  df.select_types('varchar', 'all_numeric'),
  df.exclude_types('varchar', 'all_numeric'),
  df.keys(), --- not really a star expression thing, but still useful for dynamic SQL
```

---

I'm aware that 1.1 added support for 

```sql
FROM 
  df
SELECT
  struct_pack(*COLUMNS(* ...))
```

so that some of what I suggested above can be done to the "struct of all columns in the table" already, but (a) why unnecessarily limit yourself to that particular struct? (b) I still feel it's often nicer and more powerful (because new functionality doesn't need to be combined with new syntax) to have regular methods 

---

The output type of the proposed methods depends on the inputs, so the usual bind time constant constraints will have to apply to some of its inputs.

**Comment by soerenwolfers** [comment]
`SELECT COLUMNS(struct.* EXCLUDE ..) ` etc. work, but I'd still prefer regular method syntax.

**Comment by crclark** [comment]
I'm trying to use the above `COLUMNS` trick to replace the values of fields in a struct. It works for a single-level struct, but not sure how to do it for nested structs.

This works:

```sql
WITH test AS (SELECT {'name': 'John', 'age': 30, 'city': 'NYC'} as orig_struct)
    SELECT
      orig_struct as original,
      struct_pack(*COLUMNS(orig_struct.* REPLACE ('Jane' AS name))) as updated
    FROM test;
```

This doesn't:

```sql
WITH test AS (SELECT {'person': {'name': 'John', 'age': 30}} as data)
    SELECT
      data as original,
      struct_pack(*COLUMNS(data.* REPLACE (
        struct_pack(*COLUMNS(data.person.* REPLACE (
          'Jane' AS name
        ))) AS person
      ))) as updated
    FROM test;
```

Is there a good workaround for doing deep struct field updates?

The best I can figure out is this, where I manually pull the nested struct to the top level to do the update, then push it back into the containing struct. 

The reason I am doing this is because my real struct contains a lot of fields I don't know about and don't control, so I want to avoid naming them all in my SQL.

```sql
WITH test AS (SELECT {'person': {'name': 'John', 'age': 30}, 'other': 'data', 'more': 'stuff'} as data),
  extracted_person AS (
    SELECT data, data.person as person_struct FROM test
  ),
  updated_person_fields AS (
    SELECT data, struct_pack(*COLUMNS(person_struct.* REPLACE ('Jane' AS name))) as new_person
    FROM extracted_person
  ),
  data_without_person AS (
    SELECT struct_pack(*COLUMNS(data.* EXCLUDE (person))) as data_minus_person, new_person
    FROM updated_person_fields
  )
  SELECT
    struct_insert(data_minus_person, person := new_person) as final_result
  FROM data_without_person;
```
